kart/tile/tile_dataset.py

```diff
@@ -292,14 +293,38 @@ def get_tile_summary_from_filesystem_path(self, path, *, tile_metadata=None):
     def extract_tile_metadata_from_filesystem_path(cls, path):
         raise NotImplementedError()
 
-    def diff(self, other, ds_filter=DatasetKeyFilter.MATCH_ALL, reverse=False):
+    def diff(
+        self,
+        other,
+        ds_filter=DatasetKeyFilter.MATCH_ALL,
+        reverse=False,
+        diff_format=DiffFormat.FULL,
+    ):
         """
         Generates a Diff from self -> other.
         If reverse is true, generates a diff from other -> self.
         """
+        # If the user is not asking for a tile diff, we can skip getting the diff for the feature subtree.
+        if diff_format == DiffFormat.NONE.value:
```

Comment:
I don't think this special-case is really helping you:
- if the user has requested DiffFormat.NONE, we should be able to generate the output they want without getting this far - without calling each dataset one by one and asking it to generate a diff of format DiffFormat.NONE. So in practise, this should not happen
- if this is called with DiffFormat.NONE, I think returning None is quite likely to break things, but the error would not happen here but somewhere else.

I think a better option here would be to assert that diff_format is not DiffFormat.NONE - then if this code is called with DiffFormat.NONE it will fail immediately. Calling this code with DiffFormat.NONE is not useful since it effectively asks "can I have the diff of dataset xyz but actually just return None". 

It's not necessarily bad to write a function that simply returns None in certain circumstances but I think in this case, it is much better if we make sure we don't even start the whole complicated diff code, instead of generating an "None" diff for each dataset in turn and then trying to join them all together.